---
title: "R Notebook"
output: 
---

```{r}
library(data.table)
library(tidyverse)
theme_set(theme_classic())
merged_data <- fread('ref/events_prediction.tsv')
```

# Figure 1C


```{r}
# for plot labels
dict <- data.frame(category = c('exon skipping', 'cryptic', 'intron retention', 'multi exon skipping'),
                   label = c('Exon Skipping', 'Cryptic', 'Intron Retention', 'Multi-Exon Skipping'))

# tallying event categories
events <- merged_data[in_rna_studies == 1]  %>%
  mutate(category = ifelse(!is.na(skipped_exons_count) & skipped_exons_count > 1 , 
                           'multi exon skipping', splicing_event_class),
         category = ifelse(grepl('cryptic', splicing_event_class), 'cryptic', category)) %>%
  group_by(category) %>% tally()  %>% 
  mutate(perc = n / sum(n)) %>%
  arrange(-n) %>%
  left_join(dict, by = c('category'))  %>%
  mutate(plot_category = 'mis-splicing events', fill = 1:4)

# calculating number of events per  variant
eventcounts <- merged_data[in_rna_studies == 1] %>% 
  group_by(curated_record_id) %>%
  tally(name = 'category') %>% group_by(category) %>% tally()%>%
  mutate(perc = n / sum(n),
         plot_category = '# events per variant', label = paste(category, ' events'), fill = 1:4)

# plotting
rbind(events, eventcounts) %>%
  ggplot(aes(x = perc, y = plot_category, fill = factor(fill))) + 
  geom_bar(stat = 'identity', position = position_stack(reverse=TRUE), width = 0.5) + 
  theme(axis.title.y = element_blank(), axis.title.x = element_blank(), legend.position = 'none') +
  scale_x_continuous(labels = scales::percent, expand = c(0,0)) +
  geom_text(aes(label = paste0(label, ' \nn=' ,  n)),
            position = position_stack(reverse = TRUE, vjust = 0.5), color = "white") +
  scale_fill_grey(start = 0)
ggsave('figs/1C.pdf', width = 6, height = 2)
```

# Figure 2
## Figure 2A

```{r}
# spliceAI predictions
event_deltas <- merged_data[abs(delta) >= 0.05, .(curated_record_id, splicing_event_class, 
                                                  in_rna_studies, score = delta, 
                                                  alg ='SpliceAI', skipped_exons_count)]
# MMSplice predictions
mmpred_scores <- merged_data[mmpred == 1, .(curated_record_id, splicing_event_class, 
                                            in_rna_studies, score = delta_logit_psi, 
                                            alg = 'MMSplice', skipped_exons_count)]
predictions <- rbind(event_deltas, mmpred_scores)
```

```{r}
# for plotting
lines <- data.table(alg = c(rep(c('SpliceAI'), 2), 'MMSplice'),
                    yintercept = c(0.05, 0.2, 2), 
                    splicing_event_class =c(rep(c('exon skipping'), each = 2), 'exon skipping'))

# plotting scores for all spliceAI exon skipping predictions
es_sai <- predictions %>%
  filter(splicing_event_class == 'exon skipping' & alg == 'SpliceAI') %>%
  mutate(in_rna_studies = ifelse(in_rna_studies == 0, 'not seen in RNA studies', 'seen in RNA studies'))%>%
  group_by(alg, in_rna_studies, splicing_event_class, score)%>%
  mutate(tally_id = 1:n(), grouptot = n()) %>%
  mutate(tally_id = tally_id - grouptot / 2)%>%
  mutate(in_rna_studies = factor(in_rna_studies, levels = c('seen in RNA studies', 'not seen in RNA studies'))) %>%
  ggplot(aes(x = in_rna_studies, y  = abs(score),color = in_rna_studies)) +
  geom_jitter(width =0.2, height = 0) + #facet_wrap(alg~splicing_event_class)+
  facet_wrap(~alg, nrow = 1, scales = 'free') +
  theme(strip.background = element_blank(), axis.text.x = element_text(angle = 45, hjust = 1), 
        legend.position = 'none')+
  ylab('') + xlab('') + scale_color_manual(values = c('black', 'grey'))+
  geom_hline(data = lines[alg == 'SpliceAI'][1], aes(yintercept = yintercept), linetype = 'dashed', color = 'grey') +
  geom_hline(data = lines[alg == 'SpliceAI'][2], aes(yintercept = yintercept), linetype = 'dashed') +
  scale_y_continuous(breaks = seq(0,1,0.2)) + coord_cartesian(ylim = c(0, 1))

# plotting scores for all MMSplice exon skipping predictions
es_mms <- predictions %>%
  filter(splicing_event_class == 'exon skipping' & alg == 'MMSplice') %>%
  mutate(in_rna_studies = ifelse(in_rna_studies == 0, 'not seen in RNA studies', 'seen in RNA studies'))%>%
  group_by(alg, in_rna_studies, splicing_event_class, score)%>%
  mutate(tally_id = 1:n(), grouptot = n()) %>%
  mutate(tally_id = tally_id - grouptot / 2)%>%
  mutate(in_rna_studies = factor(in_rna_studies, levels = c('seen in RNA studies', 'not seen in RNA studies'))) %>%
  ggplot(aes(x = in_rna_studies, y  = abs(score),color = in_rna_studies)) +
  geom_jitter(width =0.2, height = 0) + #facet_wrap(alg~splicing_event_class)+
  facet_wrap(~alg, nrow = 1, scales = 'free') +
  theme(strip.background = element_blank(), axis.text.x = element_text(angle = 45, hjust = 1), 
        legend.position = 'none')+
  ylab('') + xlab('') + scale_color_manual(values = c('black', 'grey'))+
  geom_hline(data = lines[alg == 'MMSplice'], aes(yintercept = yintercept), linetype = 'dashed') +
  scale_y_continuous(labels = c(0, -2, -4, -6))
es_algs <- ggarrange(es_sai, es_mms, nrow = 1)
```

```{r}
# plotting event ranks for all exon skipping 300K-RNA predictions
# plotting  events seen in RNA studies
es_krna_p <- merged_data[asj == 1 & splicing_event_class == 'exon skipping' & in_rna_studies == 1]  %>%
  mutate(in_rna_studies = ifelse(in_rna_studies == 0, 'not seen in RNA studies', 'seen in RNA studies')) %>%
  mutate(in_rna_studies = factor(in_rna_studies, levels = c('seen in RNA studies', 'not seen in RNA studies'))) %>%
  group_by(in_rna_studies, splicing_event_class, missplicing_event_rank_simple)%>%
  mutate(tally_id = 1:n(), grouptot = n()) %>%
  mutate(tally_id = tally_id - grouptot / 2)  %>%
  filter(missplicing_event_rank_simple <= 8) %>%
  ggplot(aes(x = tally_id, y = missplicing_event_rank_simple)) + 
  geom_jitter(width =0.2, height = 0.2, color = 'black')  +
  scale_x_continuous(breaks = c(0), labels = c('              seen in RNA studies')) +
  scale_y_reverse(breaks = c(seq(1,4,1), 6, 8)) +
  theme(strip.background = element_blank(),legend.position = 'none', 
        axis.text.x = element_text(angle = 45, hjust = 1), axis.title.x = element_blank(), 
        axis.title.y = element_blank()) +
  geom_hline(yintercept = 4.5, linetype = 'dashed') +
  ggtitle('300K-RNA') + coord_cartesian(ylim = c(8, 1))
# plotting  events not seen in RNA studies
es_krna_n <- merged_data[asj == 1 & splicing_event_class == 'exon skipping' & in_rna_studies == 0]  %>%
  mutate(in_rna_studies = ifelse(in_rna_studies == 0, 'not seen in RNA studies', 'seen in RNA studies')) %>%
  mutate(in_rna_studies = factor(in_rna_studies, levels = c('seen in RNA studies', 'not seen in RNA studies'))) %>%
  group_by(in_rna_studies, splicing_event_class, missplicing_event_rank_simple)%>%
  mutate(tally_id = 1:n(), grouptot = n()) %>%
  mutate(tally_id = tally_id - grouptot / 2)  %>%
  filter(missplicing_event_rank_simple <= 8) %>%
  ggplot(aes(x = tally_id, y = missplicing_event_rank_simple)) + 
  geom_jitter(width =0.2, height = 0.2, color = 'grey') +
  scale_x_continuous(breaks = c(0), labels = c('        not seen in RNA studies')) +
  scale_y_reverse(breaks = c(seq(1,4,1), 6, 8)) +
  theme(strip.background = element_blank(),legend.position = 'none', 
        axis.text.x = element_text(angle = 45, hjust = 1), axis.title.x = element_blank(), 
        axis.title.y = element_blank()) +
  geom_hline(yintercept = 4.5, linetype = 'dashed') +
  ggtitle('') + coord_cartesian(ylim = c(8, 1)) 

es_krna <- ggarrange(es_krna_p, es_krna_n)
```

```{r}
A <- ggarrange(es_krna, es_algs, nrow =1)
``` 


## Figure 2B

```{r}
# for plotting
lines <- data.table(alg = c(rep(c('SpliceAI'), 2)),
                    yintercept = c(rep(c(0.05, 0.2), 1)), 
                    splicing_event_class =c(rep(c('cryptic activation'), each = 2)))
# spliceAI scores for all cryptic activation predictions
ca_algs <- predictions %>%
  filter(grepl('cryptic', splicing_event_class) & alg == 'SpliceAI') %>%
  mutate(in_rna_studies = ifelse(in_rna_studies == 0, 'not seen in RNA studies', 'seen in RNA studies'),
         splicing_event_class = ifelse(grepl('cryptic', splicing_event_class), 
                                       'cryptic activation', splicing_event_class))%>%
  group_by(in_rna_studies, splicing_event_class, score)%>%
  mutate(tally_id = 1:n(), grouptot = n()) %>%
  mutate(tally_id = tally_id - grouptot / 2)%>%
  mutate(in_rna_studies = factor(in_rna_studies, levels = c('seen in RNA studies', 'not seen in RNA studies'))) %>%
  ggplot(aes(x = in_rna_studies, y  = abs(score),color = in_rna_studies)) +
  geom_jitter(width =0.2, height = 0) +  
  theme(strip.background = element_blank(), axis.text.x = element_text(angle = 45, hjust = 1), 
        legend.position = 'none')+
  facet_wrap(~alg, nrow = 1, scales = 'free')+
  ylab('') + xlab('') + scale_color_manual(values = c('black', 'grey'))+
  geom_hline(data = lines[alg == 'SpliceAI'][1], aes(yintercept = yintercept), linetype = 'dashed', color = 'grey') +
  geom_hline(data = lines[alg == 'SpliceAI'][2], aes(yintercept = yintercept), linetype = 'dashed') +
  scale_y_continuous(breaks = seq(0,1,0.2)) + coord_cartesian(ylim = c(0, 1))
```

```{r}
# plotting event ranks for all cryptic activation 300K-RNA predictions
# plotting  events seen in RNA studies
ca_krna_p <- merged_data[asj == 1 & grepl('cryptic', splicing_event_class) & in_rna_studies == 1]  %>%
  mutate(in_rna_studies = ifelse(in_rna_studies == 0, 'not seen in RNA studies', 'seen in RNA studies'),
         splicing_event_class = ifelse(grepl('cryptic', splicing_event_class), 
                                       'cryptic activation', splicing_event_class)) %>%
  mutate(in_rna_studies = factor(in_rna_studies, levels = c('seen in RNA studies', 'not seen in RNA studies'))) %>%
  group_by(in_rna_studies, splicing_event_class, missplicing_event_rank_simple)%>%
  mutate(tally_id = 1:n(), grouptot = n()) %>%
  mutate(tally_id = tally_id - grouptot / 2)  %>%
  filter(missplicing_event_rank_simple <= 8) %>%
  ggplot(aes(x = tally_id, y = missplicing_event_rank_simple)) + 
  geom_jitter(width =0.2, height = 0.2, color = 'black')  +
  scale_x_continuous(breaks = c(0), labels = c('              seen in RNA studies')) +
  scale_y_reverse(breaks = c(seq(1,4,1), 6, 8)) +
  theme(strip.background = element_blank(),legend.position = 'none', 
        axis.text.x = element_text(angle = 45, hjust = 1), axis.title.x = element_blank(), 
        axis.title.y = element_blank()) +
  geom_hline(yintercept = 4.5, linetype = 'dashed') +
  ggtitle('300K-RNA') + coord_cartesian(ylim = c(8, 1))

# plotting  events not seen in RNA studies
ca_krna_n <- merged_data[asj == 1 & grepl('cryptic', splicing_event_class) & in_rna_studies == 0]  %>%
  mutate(in_rna_studies = ifelse(in_rna_studies == 0, 'not seen in RNA studies', 'seen in RNA studies'),
         splicing_event_class = ifelse(grepl('cryptic', splicing_event_class), 
                                       'cryptic activation', splicing_event_class)) %>%
  mutate(in_rna_studies = factor(in_rna_studies, levels = c('seen in RNA studies', 'not seen in RNA studies'))) %>%
  group_by(in_rna_studies, splicing_event_class, missplicing_event_rank_simple)%>%
  mutate(tally_id = 1:n(), grouptot = n()) %>%
  mutate(tally_id = tally_id - grouptot / 2)  %>%
  filter(missplicing_event_rank_simple <= 8) %>%
  ggplot(aes(x = tally_id, y = missplicing_event_rank_simple)) + 
  geom_jitter(width =0.2, height = 0.2, color = 'grey')  +
  scale_x_continuous(breaks = c(0), labels = c('        not seen in RNA studies')) +
  scale_y_reverse(breaks = c(seq(1,4,1), 6, 8)) +
  theme(strip.background = element_blank(),legend.position = 'none', 
        axis.text.x = element_text(angle = 45, hjust = 1), axis.title.x = element_blank(), 
        axis.title.y = element_blank()) +
  geom_hline(yintercept = 4.5, linetype = 'dashed') +
  ggtitle('') + coord_cartesian(ylim = c(8, 1))

ca_krna <- ggarrange(ca_krna_p, ca_krna_n, nrow = 1, widths = c(0.3, 1))
```

```{r}
B <- ggarrange(ca_krna, ca_algs, nrow =1, widths = c(1.5, 1))
``` 

## Figure 2C

```{r}
# for plotting
lines <- data.table(alg = c(rep(c('SpliceAI'), 2)),
                    yintercept = c(rep(c(0.05, 0.2), 1)), 
                    splicing_event_class =c(rep(c('intron retention'), each = 2)))
# spliceAI scores for all intron retention predictions
C <- predictions %>%
  filter(splicing_event_class == 'intron retention') %>%
  mutate(in_rna_studies = ifelse(in_rna_studies == 0, 'not seen in RNA studies', 'seen in RNA studies'))%>%
  group_by(alg, in_rna_studies, splicing_event_class, score)%>%
  mutate(in_rna_studies = factor(in_rna_studies, levels = c('seen in RNA studies', 'not seen in RNA studies'))) %>%
  ggplot(aes(x = in_rna_studies, y  = abs(score),color = in_rna_studies)) +
  geom_jitter(width =0.2, height = 0) +  
  theme(strip.background = element_blank(), axis.text.x = element_text(angle = 45, hjust = 1), 
        legend.position = 'none')+
  facet_wrap(~alg, nrow = 1, scales = 'free')+
  ylab('') + xlab('') + scale_color_manual(values = c('black', 'grey'))+
  geom_hline(data = lines[alg == 'SpliceAI'][1], aes(yintercept = yintercept), linetype = 'dashed', color = 'grey') +
  geom_hline(data = lines[alg == 'SpliceAI'][2], aes(yintercept = yintercept), linetype = 'dashed') +
  scale_y_continuous(breaks = seq(0,1,0.2)) + coord_cartesian(ylim = c(0, 1))
```

## Figure 2D

```{r}
# spliceAI predictions for variants which caused 2 exon skipping/cryptic activation events
sai_pred <- merged_data[splicing_event_class != 'intron retention' & 
                          in_rna_studies == 1 & event_count_crypskip <= 2] %>%
  group_by(curated_record_id, event_count_crypskip) %>% 
  dplyr::summarise(sai_pred_count = sum(abs(delta) >= 0.2, na.rm = TRUE)) %>%
  group_by(sai_pred_count, event_count_crypskip) %>% tally() %>%
  group_by(event_count_crypskip) %>% 
  mutate(events_pred = paste0(sai_pred_count, '/' , event_count_crypskip),
         alg = 'SpliceAI', perc = n / sum(n)) %>% 
  ungroup() %>% mutate(fill = c(1,1,2,3,2)) %>% 
  select(alg, event_count_crypskip, events_pred, n, perc, fill)
# 300K-RNA predictions for variants which caused 2 exon skipping/cryptic activation events
krna_pred <- merged_data[splicing_event_class != 'intron retention' & 
              in_rna_studies == 1 & event_count_crypskip <= 2] %>%
  group_by(curated_record_id, event_count_crypskip) %>% 
  dplyr::summarise(krna_pred_count = sum(missplicing_event_rank_simple <= 4, na.rm = TRUE)) %>%
  group_by(krna_pred_count, event_count_crypskip) %>% tally() %>%
  group_by(event_count_crypskip) %>% 
  mutate(events_pred = paste0(krna_pred_count, '/' , event_count_crypskip),
         alg = '300K-RNA', perc = n / sum(n)) %>% 
  ungroup() %>% mutate(fill = c(1,2, 3, 2)) %>% 
  select(alg, event_count_crypskip, events_pred, n, perc, fill)
# plotting
D <- rbind(sai_pred, krna_pred) %>%
  mutate(fill = factor(fill, levels = c(2, 3, 1)),
         event_count_crypskip = ifelse(event_count_crypskip == 1, 'Variants with 1 mis-splicing event',
                                       'Variants with 2 mis-splicing events')) %>%
  ggplot(aes(x = alg, y = perc, fill = fill)) + 
  geom_bar(stat = 'identity', position = position_stack(reverse=TRUE))+
  facet_wrap(~event_count_crypskip)+ 
  theme(axis.title.y = element_blank(), axis.title.x = element_blank(), legend.position = 'none',
        panel.spacing = unit(2, "lines")) +
  scale_y_continuous(labels = scales::percent, expand = c(0,0)) +
  geom_text(aes(label = paste0(events_pred, ' events\nn=' ,  n)),
            position = position_stack(reverse = TRUE, vjust = 0.5), color = "black") +
  scale_fill_brewer(palette = 'Greens', direction = -1)+
  theme(plot.margin=unit(c(1,2,1,2),"cm"))

```

## Figure 2E


```{r}
# sensitivity & PPV for 300K-RNA at different cutoffs
df <- data.frame(cutoff = c(1:14))
count = c()
for (cutoff in df$cutoff) {
  count = c(count,nrow(merged_data[splicing_event_class != 'intron retention' & in_rna_studies == 1 & missplicing_event_rank_simple <= cutoff] ))
}
df$count <- count
df$sensitivity <- df$count / nrow(merged_data[splicing_event_class != 'intron retention' & in_rna_studies == 1])

# in top n events, seen / (seen + not seen)
pred_count = c()
rna_count = c()
for (cutoff in df$cutoff) {
  pred_count = c(pred_count,nrow(merged_data[splicing_event_class != 'intron retention' & missplicing_event_rank_simple <= cutoff]))
  rna_count = c(rna_count,nrow(merged_data[splicing_event_class != 'intron retention' & missplicing_event_rank_simple <= cutoff & in_rna_studies == 1]))
}
df$pred_count <- pred_count
df$rna_count <- rna_count
df$ppv <- df$rna_count / df$pred_count


name.labs <- c('Sensitivity', 'Positive Predictive Value')
names(name.labs) <- c("sensitivity", "ppv")

df_krna <- df


```

```{r}
# sensitivity & PPV for spliceAI at different cutoffs
df <- data.frame(cutoff = c(seq(0.5, 0.1, -0.1), 0.05))
count = c()
for (cutoff in df$cutoff) {
  count = c(count,nrow(merged_data[splicing_event_class != 'intron retention' & in_rna_studies == 1 & abs(delta) >= cutoff]))
}
df$count <- count

df$sensitivity <- df$count / nrow(merged_data[splicing_event_class != 'intron retention' & in_rna_studies == 1])

# above cutoff, seen / (seen + not seen)
pred_count = c()
rna_count = c()
for (cutoff in df$cutoff) {
  pred_count = c(pred_count,nrow(merged_data[splicing_event_class != 'intron retention' & abs(delta) >= cutoff]))
  rna_count = c(rna_count,nrow(merged_data[splicing_event_class != 'intron retention' & in_rna_studies == 1 & abs(delta) >= cutoff]))
}
df$pred_count <- pred_count
df$rna_count <- rna_count
df$ppv <- df$rna_count / df$pred_count



name.labs <- c('Sensitivity', 'Positive Predictive Value')
names(name.labs) <- c("sensitivity", "ppv")


df_sai <- df
```

```{r}
# plotting
E <- rbind(df_krna %>% mutate(method = '300K-RNA'),df_sai %>% mutate(method = 'SpliceAI')) %>%
  select(cutoff, sensitivity, ppv, method) %>% group_by(cutoff) %>%
  pivot_longer(cols =c('sensitivity', 'ppv')) %>%
  mutate(name = factor(name, levels = c('sensitivity', 'ppv'))) %>%
  filter(cutoff <= 7) %>%
  mutate(cutoff = factor(cutoff, levels = c(seq(0.5, 0.1, -0.1), 0.05, seq(1,7,1)))) %>%
  ggplot(aes(x = factor(cutoff), y = value, color = name) )+ geom_point() + geom_line(aes(group = name))  + 
  scale_color_manual(values = c('black','grey')) +
  theme(strip.background = element_rect(colour=NA, fill=NA),
        panel.border = element_rect(fill = NA, color = NA)) + 
  geom_text(aes(x = factor(cutoff), y = value, label = scales::percent(value, accuracy = 1)), vjust =-1) +
  coord_cartesian(ylim = c(0, 1.15)) + 
  scale_y_continuous(labels = scales::percent, expand = c(0,0)) +
  #scale_x_continuous(breaks = c(1:7), labels = function(x) paste0(intToUtf8(8804), x)) +ylab('')  +
  facet_grid(.~method, scales = 'free_x', space = "free_x") + ylab('') + xlab('') + labs(color = '')

```

# arranging & saving figure panel

```{r}
mid <- ggarrange(B,C, ncol = 2, widths = c(2.5, 1))
bottom <- ggarrange(D,E, ncol = 2, widths = c(1, 1.2))
ggarrange(A, mid, bottom,nrow = 3)
ggsave('figs/figure_two.pdf', width = 12, height = 13)
```
