---
title: "R Notebook"
output: 
---

```{r}
library(data.table)
library(tidyverse)
library(xlsx)
```

# reading in mis-splicing events seen in RNA diagnostics 

```{r}
fg_vars_events <- fread('src/fg_vars_events.tsv', sep = '\t')
```

# reading in spliceAI predictions

```{r}
sai_preds <- read.xlsx('ref/sai_preds_complete.xlsx', sheetIndex = 1)
setDT(sai_preds)

# get cryptic positions
sai_preds[, cryptic_pos := as.integer(cryptic_pos)]
sai_preds[splicing_event_class == 'cryptic donor' & strand == '+', cryptic_pos := gpos + 1]
sai_preds[splicing_event_class == 'cryptic donor' & strand == '-', cryptic_pos := gpos - 1]
sai_preds[splicing_event_class == 'cryptic acceptor' & strand == '+', cryptic_pos := gpos - 1]
sai_preds[splicing_event_class == 'cryptic acceptor' & strand == '-', cryptic_pos := gpos + 1]

# fix some transcripts
sai_preds[curated_record_id == 1, tx_id := 'ENST00000615984']
sai_preds[curated_record_id == 863, tx_id := 'ENST00000461069']
sai_preds[curated_record_id == 875, tx_id := 'ENST00000376372']
sai_preds[curated_record_id %in% c(233, 235, 943), tx_id := 'ENST00000357033']
sai_preds[curated_record_id %in% c(250, 251, 253, 254, 255, 257), tx_id := 'ENST00000397345']
sai_preds[curated_record_id %in% c(1042, 572), tx_id := 'ENST00000503292']

# fix some annotations
sai_preds[splicing_event_class == 'intron retention' &
            ir_details == 'unannotated transcript', `:=` (splicing_event_class =  "unannotated splice site hit",
                                                          ir_details = NA)]
sai_preds[splicing_event_class == 'exon skipping' &
            skipped_exons_id == 'unannotated transcript', `:=` (splicing_event_class =  "unannotated splice site hit",
                                                                skipped_exons_id = NA)]

# get relevant delta score
sai_preds[, delta := max(abs(donor), abs(acceptor))]
sai_preds[splicing_event_class == 'cryptic donor', delta := donor]
sai_preds[splicing_event_class == 'cryptic acceptor', delta := acceptor]
sai_preds[variant_type == 'donor' & splicing_event_class %in% c('exon skipping', 'intron retention'), delta := acceptor]
sai_preds[variant_type == 'acceptor' & splicing_event_class %in% c('exon skipping', 'intron retention'), delta := donor]
sai_preds[variant_type == 'donor' & grepl('mis-splicing', splicing_event_class), delta := donor]
sai_preds[variant_type == 'acceptor' & grepl('mis-splicing', splicing_event_class), delta := acceptor]
sai_preds[variant_type == 'donor' & splicing_event_class %in% c('exon skipping') &
            grepl('-', skipped_exons_id) & abs(delta) < 0.05, delta := donor]

sai_preds_filt <- sai_preds[splicing_event_class %in% c("exon skipping", 
                                                        "cryptic donor", 
                                                        "cryptic acceptor",
                                                        "intron retention") &
                              skipped_exons_id != 'alternate transcript']
```


# reading in MMSplice predictions

```{r}
mmpred_all <- fread('ref/mmsplice_predictions_all.csv')
mmpred_all <-  mmpred_all %>% 
  separate(ID, into = c('chrom', 'pos', 'ref', 'alt'), sep = ':') %>%
  separate(exons, into = c(NA, 'exons', NA), sep = ':') %>%
  separate(exons, into = c('exon_start', 'exon_end'), sep = '-', convert = TRUE) %>%
  mutate(alt= gsub("\\['|'\\]", '', alt), pos = as.integer(pos))
# annotate with curated record id
crds <- unique(fg_vars_events[, .(curated_record_id, tx_id, pos, ref, alt)])
mmpred_all <- crds[mmpred_all, on = .(tx_id =transcript_id, pos, ref, alt)][!is.na(curated_record_id)]
# annotate with splicing event class
mmpred_all[, splicing_event_class := 'exon skipping']
# annotate single-exon skipping with exon numbers
mmpred_all[, single := between(pos, exon_start - 30, exon_end + 30),by = 1:nrow(mmpred_all)]
mmpred_all[, exon_start := exon_start + 1]

ensembl <- fread('ref/ensembl_introns_exons.tsv.gz')
mmpred_single <- ensembl[region_type == 'exon', 
                         .(tx_id, 
                           exon_start = region_start, 
                           exon_end = region_end, 
                           skipped_exons_id = region_no)][mmpred_all[single == TRUE], 
                                                          on= .(tx_id, exon_start, exon_end)]


# annotate multi-exon skipping
mmpred_multi <- mmpred_all[single == FALSE]
mmpred_multi[, skipped_exons_id := c('2-3', '16-17', '3-4', '4-5', '9-10')]

mmpred_pred <- rbind(mmpred_single, mmpred_multi)
```

# Reading in 300K-RNA predictions

```{r}
msd_filt <- fread('src/300KRNA_filt.tsv.gz', nThread = 4)

# calculate event ranks
msd_filt[splicing_event_class != 'normal splicing', missplicing_event_rank := rowid(splice_site_pos), by = list(tx_id)]
msd_filt[splicing_event_class != 'normal splicing' & (skipped_exons_count <= 2 |abs(cryptic_distance) <= 600),
    missplicing_event_rank_simple := rowid(splice_site_pos), by = list(tx_id)]

#alter splicing event class to line up with how mis-splicing events are annotated
msd_filt[, splicing_event_class_save := splicing_event_class]
msd_filt[splicing_event_class_save == 'alternative donor (annotated)', splicing_event_class := 'cryptic donor']
msd_filt[splicing_event_class_save == 'alternative acceptor (annotated)', splicing_event_class := 'cryptic acceptor']
msd_filt[splicing_event_class_save == 'exon skipping (annotated)', splicing_event_class := 'exon skipping']

#make cryptic pos column
msd_filt[ss_type == 'donor' & splicing_event_class == 'cryptic donor', cryptic_pos := donor_pos]
msd_filt[ss_type == 'acceptor' & splicing_event_class == 'cryptic acceptor', cryptic_pos := acceptor_pos]

# select relevant columns
msd_filt_join <- msd_filt[, .(tx_id, splice_site_pos, splicing_event_class, skipped_exons_id,
                              cryptic_pos, event_rank, missplicing_event_rank, missplicing_event_rank_simple,
                              in_gtex, in_sra, missplicing_inframe,gtex_sample_count,
                              max_uniq_reads, sra_sample_count, sample_count, cryptic_distance,skipped_exons_count)]


# get normal splicing sample count
ns_sc <- msd_filt_join[splicing_event_class == 'normal splicing', .(tx_id, splice_site_pos,
                                                                    sample_count_ns =sample_count,
                                                                    ns_reads = max_uniq_reads)]
msd_filt_join <- ns_sc[msd_filt_join, on = .(tx_id, splice_site_pos)]

# filter to relevant events
msd_filt_join <- msd_filt_join[skipped_exons_count <= 2 | abs(cryptic_distance) <= 600]

# annotate with curated record id
crds <- unique(fg_vars_events[, .(curated_record_id, tx_id, splice_site_pos)])
msd_filt_join <- crds[msd_filt_join, on = .(tx_id, splice_site_pos)]
```

# making master dataset with events and all predictions for the purpose of making figures

```{r}
# join on CRD, splicing_event_class, cryptic_pos, skipped_exons_id
sai <- sai_preds_filt[, .(curated_record_id, splicing_event_class, cryptic_pos, skipped_exons_id, delta, sai_pred = 1)]
asj <- msd_filt_join[, .(curated_record_id, splicing_event_class, cryptic_pos, cryptic_distance, 
                         skipped_exons_id,skipped_exons_count,
                         missplicing_event_rank_simple, sample_count, sample_count_ns, asj = 1)]
asj[missplicing_event_rank_simple <= 4, asj_top4 := 1]
mmpred <- mmpred_pred[, .(curated_record_id,tx_id,splicing_event_class, skipped_exons_id, delta_logit_psi, cryptic_pos = NA)]
mmpred[, mmpred := 1]
mmpred[delta_logit_psi <= -2, mmpred_es := 1]
fg <- fg_vars_events[, .(curated_record_id, splicing_event_class, cryptic_pos, skipped_exons_id, cryptic_modified_event_flag, in_rna_studies = 1)]


fg_sai <- merge(fg, sai, by = c('curated_record_id', 'splicing_event_class', 'cryptic_pos', 'skipped_exons_id'), all = T)
fg_sai_mmpred <- merge(fg_sai, mmpred, by = c('curated_record_id', 'splicing_event_class', 'cryptic_pos', 'skipped_exons_id'), all = T)
merged_data <- merge(fg_sai_mmpred, asj, by = c('curated_record_id', 'splicing_event_class', 'cryptic_pos', 'skipped_exons_id'), all = T)

merged_data[is.na(in_rna_studies), in_rna_studies := 0]
merged_data[is.na(sai_pred), sai_pred := 0]
merged_data[is.na(asj_top4), asj_top4 := 0]
merged_data[is.na(mmpred), mmpred := 0]
merged_data[is.na(mmpred_es), mmpred_es := 0]
merged_data[is.na(asj), asj := 0]

# annotate with event count, variant type
info <- unique(fg_vars_events[, .(curated_record_id, event_count, event_count_crypskip, variant_type, cryptic_modified_variant_flag)])
merged_data <- info[merged_data, on = .(curated_record_id)]

fwrite(merged_data, 'ref/events_prediction.tsv', sep = '\t')
```

