---
title: "R Notebook"
output: 
---


```{r}
library(data.table)
library(tidyverse)
theme_set(theme_classic())
library(openxlsx)
wb <- createWorkbook()
```

```{r}
merged_data <- fread('../ref/vars_events_prediction.tsv')
```


# Figure S1: PR curve of our 88 variants, 300K-RNA predictions with different ranking methods
```{r}
pr_data_sv <- merged_data[variant_category == 'ours' & 
                            splicing_event_class != 'intron retention' &
                         (splicevault == 1 | in_rna_studies == 1)]

cutoffs <- data.table(splicevault_rank_cutoffs = 1:25)

# calculating sensitivity & PPV for different ranking methods
# max double exon skipping, cryptics 600 nt away
cutoffs[, sv_TP := unlist(lapply(splicevault_rank_cutoffs, 
                                 function(x) nrow(pr_data_sv[in_rna_studies == 1 & missplicing_event_rank_simple <= x])))]
cutoffs[, sv_FP := unlist(lapply(splicevault_rank_cutoffs, 
                                 function(x) nrow(pr_data_sv[in_rna_studies == 0 & missplicing_event_rank_simple <= x])))]
cutoffs[, sv_FN := unlist(lapply(splicevault_rank_cutoffs, 
                                 function(x) nrow(pr_data_sv[in_rna_studies == 1 & (missplicing_event_rank_simple > x |
                                                                is.na(missplicing_event_rank_simple))])))]
cutoffs[, sv_sens := sv_TP / (sv_TP + sv_FN)]
cutoffs[, sv_ppv := sv_TP / (sv_TP + sv_FP)]



# max single exon skipping, cryptics 250 nt away
cutoffs[, svbasic_TP := unlist(lapply(splicevault_rank_cutoffs, 
                                 function(x) nrow(pr_data_sv[in_rna_studies == 1 & missplicing_event_rank_basic <= x])))]
cutoffs[, svbasic_FP := unlist(lapply(splicevault_rank_cutoffs, 
                                 function(x) nrow(pr_data_sv[in_rna_studies == 0 & missplicing_event_rank_basic <= x])))]
cutoffs[, svbasic_FN := unlist(lapply(splicevault_rank_cutoffs, 
                                 function(x) nrow(pr_data_sv[in_rna_studies == 1 & (missplicing_event_rank_basic > x |
                                                                is.na(missplicing_event_rank_basic))])))]
cutoffs[, svbasic_sens := svbasic_TP / (svbasic_TP + svbasic_FN)]
cutoffs[, svbasic_ppv := svbasic_TP / (svbasic_TP + svbasic_FP)]

# all missplicing events
cutoffs[, svall_TP := unlist(lapply(splicevault_rank_cutoffs, 
                                 function(x) nrow(pr_data_sv[in_rna_studies == 1 & missplicing_event_rank <= x])))]
cutoffs[, svall_FP := unlist(lapply(splicevault_rank_cutoffs, 
                                 function(x) nrow(pr_data_sv[in_rna_studies == 0 & missplicing_event_rank <= x])))]
cutoffs[, svall_FN := unlist(lapply(splicevault_rank_cutoffs, 
                                 function(x) nrow(pr_data_sv[in_rna_studies == 1 & (missplicing_event_rank > x |
                                                                is.na(missplicing_event_rank))])))]
cutoffs[, svall_sens := svall_TP / (svall_TP + svall_FN)]
cutoffs[, svall_ppv := svall_TP / (svall_TP + svall_FP)]

# all missplicing events, sample count > 10
cutoffs[, svall10_TP := unlist(lapply(splicevault_rank_cutoffs, 
                                 function(x) nrow(pr_data_sv[in_rna_studies == 1 & missplicing_event_rank <= x & sample_count > 10])))]
cutoffs[, svall10_FP := unlist(lapply(splicevault_rank_cutoffs, 
                                 function(x) nrow(pr_data_sv[in_rna_studies == 0 & missplicing_event_rank <= x & sample_count > 10])))]
cutoffs[, svall10_FN := unlist(lapply(splicevault_rank_cutoffs, 
                                 function(x) nrow(pr_data_sv[in_rna_studies == 1 & (missplicing_event_rank > x |
                                                                is.na(missplicing_event_rank) | sample_count <= 10)])))]
cutoffs[, svall10_sens := svall10_TP / (svall10_TP + svall10_FN)]
cutoffs[, svall10_ppv := svall10_TP / (svall10_TP + svall10_FP)]
```

```{r}
# plot PR curve
plot_df <- cutoffs %>%
  dplyr::select(-contains('TP'), -contains('FP'), -contains('FN')) %>%
  pivot_longer(cols = c('sv_sens', 'sv_ppv', 'svbasic_sens', 'svbasic_ppv', 'svall_sens', 'svall_ppv', 'svall10_sens', 'svall10_ppv')) %>%
  separate(name, into = c('alg', 'metric'), sep = '_') %>%
  pivot_wider(names_from = metric, values_from = value) %>%
  mutate(alg = factor(alg, levels = c('sv', 'svbasic', 'svall10', 'svall')))
setDT(plot_df)

ggplot(plot_df, aes(x = sens, y = ppv, color = alg)) +
  geom_point() + geom_path() +
  coord_cartesian(xlim = c(0.4,1), y = c(0,1)) +
  xlab('Sensitivity') + ylab('Positive Predictive Value') +
  labs(color = '')+
  scale_color_nejm(labels = c('max 2 exons skipped, cryptics within 600 nt',
                              'max 1 exon skipped, cryptics within 250 nt',
                              'all events seen in > 10 samples',
                              'all events')) +
  scale_x_continuous(breaks = seq(0,1,0.1))+
  scale_y_continuous(breaks = seq(0,1,0.1)) +
  theme(panel.grid.major = element_line(colour = "lightgrey", size = 0.1)) +
  geom_text(data=plot_df[alg == 'sv'][c(1:4)],aes(x=sens,y=ppv + 0.05,label=splicevault_rank_cutoffs), color='#ae4532')+
  geom_text(data=plot_df[alg == 'svall'][c(1:4)],aes(x=sens,y=ppv - 0.05,label=splicevault_rank_cutoffs), color='#418454')+
  geom_text(data=plot_df[alg == 'svall10'][c(1:4)],aes(x=sens,y=ppv + 0.05,label=splicevault_rank_cutoffs), color='#d68a3d')+
  geom_text(data=plot_df[alg == 'svbasic'][c(1:4)],aes(x=sens,y=ppv - 0.05,label=splicevault_rank_cutoffs), color='#2e76b3')

ggsave('../figs/S1.pdf', height = 5, width = 8)

addWorksheet(wb, "figure_E1")
writeData(wb, "figure_E1", as.data.frame(plot_df), startRow = 1, startCol = 1)
saveWorkbook(wb, file = "../figs/source_data/source_data_extended_data_fig_1.xlsx", overwrite = TRUE)

```


