---
title: "R Notebook"
output: 
---

```{r}
library(data.table)
library(tidyverse)
library(openxlsx)
theme_set(theme_classic())
wb <- createWorkbook()
```

```{r}
merged_data <- fread('../ref/vars_events_prediction.tsv')
merged_data %>%
  dplyr::select(curated_record_id, variant_category) %>% 
  distinct() %>% group_by(variant_category) %>% tally()
```


# Figure 2D: % of mis-splicing events seen in 300K-RNA at different event ranks

```{r}
df_2a <- merged_data[splicing_event_class != 'intron retention' & 
                         variant_category == 'ours' &
                         (in_rna_studies == 1)]

plot_data <- data.table(rank_cutoff = 1:25)
plot_data[, count_in_sv := sapply(rank_cutoff, function(x) nrow(df_2a[missplicing_event_rank_simple <= x]))]
plot_data[, prop_in_sv := count_in_sv / nrow(df_2a)]

D = plot_data  %>%
  filter(rank_cutoff <= 10) %>%
  ggplot(aes(x = factor(rank_cutoff), y = prop_in_sv)) + 
  geom_point(color = '#2e72af') + geom_path(group = 1, color = '#2e72af') + 
  geom_text(aes(x = factor(rank_cutoff), y = prop_in_sv, 
                label = scales::percent(prop_in_sv, accuracy = 1)), vjust =-1, color = '#2e72af', size = 3) +
  coord_cartesian(ylim = c(0.5, 1.05)) + 
  scale_y_continuous(labels = scales::percent, expand = c(0,0)) + 
  ylab('Detected events (%)') + xlab('Event Rank in 300K-RNA') + labs(color = '') +
  scale_x_discrete(labels = function(x) paste0(intToUtf8(8804), x)) +
  theme(plot.margin = margin(1,1,1.5,1.2, "cm"))

ggsave('../figs/2D.pdf', width = 4, height = 3.5)

addWorksheet(wb, "figure_2D")
writeData(wb, "figure_2D", as.data.frame(plot_data  %>% filter(rank_cutoff <= 10) ), startRow = 1, startCol = 1)

```



# Figure 2E: proportion of TPs detected in random subsamples of 300K-RNA

```{r}
all_samples_unnest <- fread('../ref/allSamples_snapcount_unnested.tsv.gz')
```

```{r}
merged_data <- fread('../ref/vars_events_prediction.tsv')

df_2 <- merged_data[splicing_event_class != 'intron retention' & 
                         variant_category == 'ours' &
                         (in_rna_studies == 1 | missplicing_event_rank_simple <= 4)]
splicevault <- fread('../src/300KRNA_filt.tsv.gz')
df_2 = splicevault[df_2, on = c(intersect(names(df_2), names(splicevault)))]

```

```{r}
df_2[strand == '+', `:=` (start = donor_pos, end = acceptor_pos)]
df_2[strand == '-', `:=` (start = acceptor_pos, end = donor_pos)]
all_samples_unnest[, chr := gsub('chr', '', chromosome)]

all_samples_unnest = all_samples_unnest[df_2, on = .(chr, start, end)]
```

```{r}
TPs_detected = all_samples_unnest %>%
  filter(in_rna_studies == 1) %>%
  select(sample, curated_record_id, missplicing_event_rank_simple) %>%
  mutate(TP_detected = paste(curated_record_id, missplicing_event_rank_simple, sep = ':')) %>%
  select(sample, TP_detected) 
  
# sample read counts
sampleinfo_gtex = fread('../ref/samples_gtexv2.tsv.gz')
sampleinfo_sra = fread('../ref/samples_srav3h.tsv.gz')

sample_info = rbind(sampleinfo_gtex[, .(sample = rail_id,star.all_mapped_reads, info = SMTS)],
                           sampleinfo_sra[, .(sample = rail_id,star.all_mapped_reads, info = study_acc)])

random_sample_sens = data.table(sample_counts = c(10^(seq(1,6,0.2))[1:23], nrow(sample_info)))

TP_count = c()
for (x in 1:nrow(random_sample_sens)) {
  sample_count = random_sample_sens$sample_counts[x]
  for (i in 1:20) {
    TP_count = c(TP_count, length(unique(TPs_detected[sample_info[sample(nrow(sample_info), 
                                                                         sample_count), ], 
                                                      on = .(sample)][!is.na(TP_detected), TP_detected])))
  }
}

random_sample_sens = random_sample_sens %>% slice(rep(1:n(), each = 20)) %>% mutate(TP_counts = TP_count, random_sample = rep(1:20, nrow(random_sample_sens)))



E = random_sample_sens %>% ggplot(aes(x = sample_counts, y = TP_counts/119 )) + 
  geom_point(alpha = 0.5, color = 'grey',size = 0.4) + scale_x_log10() + geom_smooth(color = '#2e72af') +
  xlab('Size of random sample of 300K-RNA') + ylab('Detected Events (%)') +
  scale_y_continuous(labels = scales::percent)



addWorksheet(wb, "figure_2E")
writeData(wb, "figure_2E", as.data.frame(random_sample_sens ), startRow = 1, startCol = 1)
```

# 2F: overlap between samples where top-1 and top-2 are seen

```{r}
props_overlap = all_samples_unnest  %>%
  select(sample, reads, curated_record_id, missplicing_event_rank_simple) %>%
  filter(missplicing_event_rank_simple <= 2) %>%
  mutate(TP_detected = paste(curated_record_id, missplicing_event_rank_simple, sep = ':'))%>% 
  select(sample, curated_record_id, missplicing_event_rank_simple, reads) %>% 
  pivot_wider(values_from = reads, names_from = missplicing_event_rank_simple, names_prefix = 'event_') %>%
  mutate(event_1 = ifelse(is.na(event_1), 0, event_1),
         event_2 = ifelse(is.na(event_2), 0, event_2)) %>%
  group_by(curated_record_id) %>%
  dplyr::summarise(samples_event_1 = length(sample[event_1 > 0 & event_2 == 0]),
                   samples_event_2 = length(sample[event_1 == 0 & event_2 > 0]),
                   samples_both = length(sample[event_1 > 0 & event_2 > 0]), .groups ='keep') %>%
  pivot_longer(cols = contains('samples_')) %>%
  mutate(total = sum(value), prop = value / total,
         name = factor(name, levels = c('samples_event_1', 'samples_event_2', 'samples_both')) )


panelF = props_overlap  %>%
  ggplot(aes(x = name, y= prop)) + 
  geom_boxplot(outlier.alpha = 0) + geom_jitter(color = '#2e72af' , size = 0.4) +
  xlab('Event(s) detected') + ylab('Percent of samples') +
  scale_x_discrete(labels = c('Only Top-1', 'Only Top-2', 'Top-1 and Top-2'))  + 
  scale_y_continuous(labels = scales::percent) 


addWorksheet(wb, "figure_2F")
writeData(wb, "figure_2F", as.data.frame(props_overlap ), startRow = 1, startCol = 1)

```
```{r}
saveWorkbook(wb, file = "../figs/source_data/source_data_fig_2.xlsx", overwrite = TRUE)

```

```{r}
ggarrange(E, panelF, nrow = 1, widths = c(1.5, 1))
ggsave('../figs/2EF.pdf', width = 6, height = 2.5)
```

