---
title: "R Notebook"
output: 
---


```{r}
library(data.table)
library(tidyverse)
theme_set(theme_classic())
options(scipen=999)
library(ggsci)
library(ggpubr)
library(scales)
library(metR)
library(openxlsx)
wb <- createWorkbook()
```



```{r}
merged_data <- fread('../ref/vars_events_prediction.tsv')
```


# Figure 4D: Example SAI deltas plot with transcript

```{r}
case_numbers <- unique(merged_data[variant_category == 'ours', .(curated_record_id)]) %>%
  mutate(case_number = 1:88)

preds <- fread('../ref/sai_preds_complete.csv')
preds <- unique(merged_data[, .(curated_record_id, variant_category)])[preds, on = .(curated_record_id)]
preds <- case_numbers[preds, on = .(curated_record_id)]
preds <- preds[variant_category == 'ours']

merged_data[in_rna_studies == 1, .(curated_record_id, splicing_event_class, 
                                cryptic_pos, skipped_exons_id, 
                                in_rna_studies)]

preds <- unique(merged_data[, .(curated_record_id, splicing_event_class, 
                                cryptic_pos, skipped_exons_id, 
                                in_rna_studies)])[preds, on = .(curated_record_id, splicing_event_class, 
                                cryptic_pos, skipped_exons_id)]

preds[splicing_event_class == 'mis-splicing', in_rna_studies := 2]
preds[, in_rna_studies := factor(in_rna_studies, levels = c(0,1,2))]

```


```{r}
plot_dist = 150

pos = preds[case_number == 6 & abs(dist_from_variant) < plot_dist & delta > 0] %>%
  ggplot(aes(fill = in_rna_studies)) + 
   geom_rect(aes(xmax=as.numeric(dist_from_variant)-.25, xmin=as.numeric(dist_from_variant)+.25, 
       ymax=delta, ymin=0.001)) + 
   scale_y_log10(expand = c(0,0)) +
  geom_hline(yintercept = c(0.011), linetype = 'dotted', size = 0.5) + 
  coord_cartesian(xlim = c(-150, 50)) + 
  scale_fill_manual(values = c('grey','#ae4631'))

neg = preds[case_number == 6 & abs(dist_from_variant) < plot_dist & delta < 0] %>%
  ggplot(aes(fill = in_rna_studies)) + 
   geom_rect(aes(xmax=-as.numeric(dist_from_variant)-.25, xmin=-as.numeric(dist_from_variant)+.25, 
       ymax=abs(delta), ymin=0.001)) + 
  geom_hline(yintercept = c(0.011), linetype = 'dotted', size = 0.5)+ 
  coord_cartesian(xlim = c(-150, 50)) + 
  scale_y_continuous(trans=reverselog_trans(base=10),
                     labels=trans_format("identity", function(x) -x), expand = c(0,0))+
  scale_x_continuous(position = "top")  + 
  scale_fill_manual(values = c('grey','#ae4631', 'black'))

tx = ggplot(data.frame(x = c(-150, 50), y = c(-1, 1)), aes(x, y)) +
  annotate("rect", xmin = -123, xmax = 5, ymin = -1, ymax = 1,
           alpha = 1,fill = "black") +
  coord_cartesian(xlim = c(-150, 50)) +
  geom_hline(yintercept = 0) + 
  theme_void()

D = ggarrange(pos,tx, neg,ncol = 1, common.legend = TRUE, heights = c(1, 0.2, 1))




addWorksheet(wb, "figure_4D")
writeData(wb, "figure_4D", as.data.frame(preds[case_number == 6 & abs(dist_from_variant) < plot_dist ]), startRow = 1, startCol = 1)
```


# Figure 4E: PR curve of our 88 variants, SpliceAI & 300K-RNA predictions of exon-skipping & cryptic-activation

```{r}
# Remove insertion-deletions from comparative PR curve. 86 variants in total.
# keep only cryptic and exon skipping predictions form spliceAI (remove other / uninterpretable predictions)
pr_data <- merged_data[!curated_record_id %in% c(871, 913, 927) &
                         splicing_event_class %in% c('cryptic acceptor', 'cryptic donor', 'exon skipping') &
                         variant_category == 'ours' &
                         (spliceai == 1 | splicevault == 1 | in_rna_studies == 1)] 

# #find thresholds where SAI predicts the same number of events as 300K-RNA top-1, 2, 3, 4, etc.
# sai_cutoffs <- data.table(cutoffs = seq(0, 1, 0.000001))
# sai_cutoffs[, pred_count := sapply(cutoffs, function(x) nrow(pr_data[spliceai == 1 & abs(delta) >= x]))]
# fwrite(sai_cutoffs, '../ref/sai_cutoffs.csv')
sai_cutoffs <- fread('../ref/sai_cutoffs.csv')

cutoffs <- data.table(splicevault_rank_cutoffs = 1:25)
cutoffs[, pred_count := sapply(splicevault_rank_cutoffs, function(x) nrow(pr_data[missplicing_event_rank_simple <= x]))]

cutoffs[pred_count <= nrow(pr_data[spliceai == 1]), sai_threshold := sapply(pred_count, function(x)  min(sai_cutoffs[pred_count == x,cutoffs]))]
cutoffs[pred_count > nrow(pr_data[spliceai == 1]), sai_threshold := NA]

```

```{r}
# calculating sensitivity & PPV
cutoffs[, sv_TP := unlist(lapply(splicevault_rank_cutoffs, 
                                 function(x) nrow(pr_data[in_rna_studies == 1 & missplicing_event_rank_simple <= x])))]
cutoffs[, sv_FP := unlist(lapply(splicevault_rank_cutoffs, 
                                 function(x) nrow(pr_data[in_rna_studies == 0 & missplicing_event_rank_simple <= x])))]
cutoffs[, sv_FN := unlist(lapply(splicevault_rank_cutoffs, 
                                 function(x) nrow(pr_data[in_rna_studies == 1 & (missplicing_event_rank_simple > x |
                                                                is.na(missplicing_event_rank_simple))])))]
cutoffs[, sv_sens := sv_TP / (sv_TP + sv_FN)]
cutoffs[, sv_ppv := sv_TP / (sv_TP + sv_FP)]

cutoffs[, sai_TP := unlist(lapply(sai_threshold, 
                                 function(x) nrow(pr_data[in_rna_studies == 1 & abs(delta) >= x])))]
cutoffs[, sai_FP := unlist(lapply(sai_threshold, 
                                 function(x) nrow(pr_data[in_rna_studies == 0 & abs(delta) >= x])))]
cutoffs[, sai_FN := unlist(lapply(sai_threshold, 
                                 function(x) nrow(pr_data[in_rna_studies == 1 & (abs(delta) < x | is.na(delta))])))]
cutoffs[, sai_sens := sai_TP / (sai_TP + sai_FN)]
cutoffs[, sai_ppv := sai_TP / (sai_TP + sai_FP)]
```



```{r}
# plot PR curve
plot_df <- cutoffs %>%
  filter(splicevault_rank_cutoffs %in% c(1:5, seq(6, 10, 2), 16)) %>%
  dplyr::select(-sv_TP, -sv_FP, -sv_FN, -sai_TP, -sai_FP, -sai_FN) %>%
  pivot_longer(cols = c('sv_sens', 'sv_ppv', 'sai_sens', 'sai_ppv')) %>%
  separate(name, into = c('alg', 'metric'), sep = '_') %>%
  pivot_wider(names_from = metric, values_from = value)
setDT(plot_df)

E <- ggplot(plot_df[!is.na(sai_threshold)], aes(x = sens, y = ppv, color = alg)) +
  geom_point() + geom_path() +
  coord_cartesian(xlim = c(0.5,1), y = c(0,1)) +
  xlab('Sensitivity') + ylab('Positive Predictive Value') +
  labs(color = '')+
  scale_color_nejm(labels = c("SpliceAI", "300K-RNA"))+
  geom_text(data=plot_df[alg == 'sv'][c(1:4)],aes(x=sens,y=ppv + 0.08,label=splicevault_rank_cutoffs), color='#2e72af', size = 3)+
  geom_text(data=plot_df[alg == 'sai'][c(1:4)],aes(x=sens - 0.04,y=ppv - 0.05,label=round(sai_threshold,3)), color='#ae4631', size = 3) +
  scale_x_continuous(breaks = seq(0,1,0.1))+
  scale_y_continuous(breaks = seq(0,1,0.1)) +
  theme(panel.grid.major = element_line(colour = "lightgrey", size = 0.1))+
  theme(plot.margin = margin(1,2,1.5,2, "cm"))



addWorksheet(wb, "figure_4E")
writeData(wb, "figure_4E", as.data.frame(plot_df[!is.na(sai_threshold)]), startRow = 1, startCol = 1)

```


# Figure 2F: TP / FP dot plots for exon skipping

#spliceai & splicevault delta jitter plot

```{r}
jitter_data_sv <- merged_data[variant_category == 'ours' & 
                                !curated_record_id %in% c(871, 913, 927) &
                            splicing_event_class != 'intron retention' &
                         (!is.na(missplicing_event_rank_simple) | in_rna_studies == 1)] %>%
  mutate(in_rna_studies = ifelse(in_rna_studies == 0, 'not seen in RNA studies', 'seen in RNA studies')) %>%
  mutate(in_rna_studies = factor(in_rna_studies, levels = c('seen in RNA studies', 'not seen in RNA studies')))
jitter_data_sai <- merged_data[variant_category == 'ours' & 
                                 !curated_record_id %in% c(871, 913, 927) &
                            splicing_event_class != 'intron retention' &
                         (spliceai == 1 | in_rna_studies == 1)]%>%
  mutate(in_rna_studies = ifelse(in_rna_studies == 0, 'not seen in RNA studies', 'seen in RNA studies')) %>%
  mutate(in_rna_studies = factor(in_rna_studies, levels = c('seen in RNA studies', 'not seen in RNA studies')))

setDT(jitter_data_sv)
jitter_data_sv[, rank_above_4 := ifelse(missplicing_event_rank_simple <= 4, 1,0)]
labs_sv <- jitter_data_sv  %>% 
  mutate(splicing_event_class = ifelse(grepl('cryptic', splicing_event_class), 'cryptic', splicing_event_class)) %>% 
  group_by(splicing_event_class, rank_above_4, in_rna_studies) %>% tally() 
setDT(labs_sv)

setDT(jitter_data_sai)
jitter_data_sai[, delta_above_thresh := ifelse(abs(delta) < cutoffs[splicevault_rank_cutoffs == 4, sai_threshold] |
                                                 is.na(delta),0,1)]
labs_sai <- jitter_data_sai %>% 
  mutate(splicing_event_class = ifelse(grepl('cryptic', splicing_event_class), 'cryptic', splicing_event_class)) %>%
  group_by(splicing_event_class, delta_above_thresh, in_rna_studies) %>% tally() 
setDT(labs_sai)
```

```{r}
es_sai <- jitter_data_sai %>%
  filter(splicing_event_class == 'exon skipping') %>%
  mutate(color = ifelse(is.na(delta), 'not seen', as.character(in_rna_studies))) %>%
  mutate(delta = ifelse(is.na(delta), 0.001, delta))  %>%
  ggplot(aes(x = in_rna_studies, y  = abs(delta),color = color)) +
  geom_jitter(width =0.2, height = 0, size = 0.5, alpha =0.4) +
  theme(strip.background = element_blank(), axis.text.x = element_text(angle = 45, hjust = 1), 
        legend.position = 'none')+
  ylab('\u0394-score') + xlab('') + 
  scale_color_manual(values = c('black', 'grey','#ae4631'))+
  scale_y_log10() +
  geom_hline(yintercept = 0.011562, linetype = 'dashed') + 
  ggtitle('SpliceAI')+
  theme(plot.margin = margin(0.5,0.5,0.5,0.5, "cm"))+
  annotate("text", x = 1, y = 0.017, label = paste('TP =', labs_sai[splicing_event_class == 'exon skipping' & 
                                                                     !grepl('not', in_rna_studies) &
                                                                     delta_above_thresh == 1 , n]), size = 3) +
  annotate("text", x = 1, y = 0.008, label = paste('FN =', labs_sai[splicing_event_class == 'exon skipping' & 
                                                                     !grepl('not', in_rna_studies) &
                                                                     delta_above_thresh == 0 , n]), size = 3)+
  annotate("text", x = 2, y = 0.017, label = paste('FP =',labs_sai[splicing_event_class == 'exon skipping' & 
                                                                    grepl('not', in_rna_studies) &
                                                                    delta_above_thresh == 1 , n]), size = 3)

es_sv <- jitter_data_sv %>%
  filter(splicing_event_class == 'exon skipping')%>%
  ggplot(aes(x = factor(in_rna_studies), y = missplicing_event_rank_simple, color = in_rna_studies)) + 
  geom_jitter(width =0.2, height = 0.1, size = 0.5, alpha =0.4) + 
  scale_color_manual(values = c('#2e72af', 'grey')) +
  scale_y_reverse(breaks = seq(1,25,2)) +
  theme(strip.background = element_blank(),legend.position = 'none', 
        axis.text.x = element_text(angle = 45, hjust = 1), axis.title.x = element_blank()) +
  geom_hline(yintercept = 4.2, linetype = 'dashed') +
  ggtitle('300K-RNA')  +
  ylab('event rank')+
  theme(plot.margin = margin(0.5,0.5,0.5,0.5, "cm"))+
  annotate("text", x = 1, y = 3.5, label = paste('TP =', labs_sv[splicing_event_class == 'exon skipping' & 
                                                                     !grepl('not', in_rna_studies) &
                                                                     rank_above_4 == 1 , n]), size = 3) +
  annotate("text", x = 1, y = 4.8, label = paste('FN =', labs_sv[splicing_event_class == 'exon skipping' & 
                                                                     !grepl('not', in_rna_studies) &
                                                                     rank_above_4 == 0 , n]), size = 3)+
  annotate("text", x = 2, y = 3.5, label = paste('FP =',labs_sv[splicing_event_class == 'exon skipping' & 
                                                                    grepl('not', in_rna_studies) &
                                                                    rank_above_4 == 1 , n]), size = 3)



panelF <- ggarrange(es_sv, es_sai, nrow = 1)  %>% annotate_figure(fig.lab = 'Exon Skipping', fig.lab.size = 12)

sv = jitter_data_sv %>%
  filter(splicing_event_class == 'exon skipping') %>%
  select(in_rna_studies, score = missplicing_event_rank_simple, meets_threshold = rank_above_4)%>%
  mutate(alg ='splicevault')

sai = jitter_data_sai %>%
  filter(splicing_event_class == 'exon skipping')%>%
  select(in_rna_studies, score = delta, meets_threshold = delta_above_thresh)%>%
  mutate(alg ='spliceai')

addWorksheet(wb, "figure_4F")
writeData(wb, "figure_4F", as.data.frame(rbind(sv, sai)), startRow = 1, startCol = 1)
```

# Figure 2G: TP / FP dot plots for cryptic activation

```{r}
ca_sai <- jitter_data_sai %>%
  filter(grepl('cryptic', splicing_event_class)) %>%
  mutate(color = ifelse(is.na(delta), 'not seen', as.character(in_rna_studies))) %>%
  mutate(delta = ifelse(is.na(delta), 0.001, delta))    %>%
  ggplot(aes(x = in_rna_studies, y  = abs(delta),color = color)) +
  geom_jitter(width =0.2, height = 0, size = 0.5, alpha =0.4) +
  theme(strip.background = element_blank(), axis.text.x = element_text(angle = 45, hjust = 1), 
        legend.position = 'none')+
  ylab('\u0394-score') + xlab('') + scale_color_manual(values = c('black', 'grey', '#ae4631'))+
  scale_y_log10() +
  geom_hline(yintercept = 0.011562, linetype = 'dashed') + 
  ggtitle('SpliceAI')+
  theme(plot.margin = margin(0.5,0.5,0.5,0.5, "cm"))+
  annotate("text", x = 1, y = 0.017, label = paste('TP =', labs_sai[grepl('cryptic', splicing_event_class) & 
                                                                     !grepl('not', in_rna_studies) &
                                                                     delta_above_thresh == 1 , n]), size = 3) +
  annotate("text", x = 1, y = 0.008, label = paste('FN =', labs_sai[grepl('cryptic', splicing_event_class) & 
                                                                     !grepl('not', in_rna_studies) &
                                                                     delta_above_thresh == 0 , n]), size = 3)+
  annotate("text", x = 2, y = 0.017, label = paste('FP =',labs_sai[grepl('cryptic', splicing_event_class) & 
                                                                    grepl('not', in_rna_studies) &
                                                                    delta_above_thresh == 1 , n]), size = 3)

ca_sv <- jitter_data_sv %>%
  filter(grepl('cryptic', splicing_event_class))%>%
  ggplot(aes(x = factor(in_rna_studies), y = missplicing_event_rank_simple, color = in_rna_studies)) + 
  geom_jitter(width =0.2, height = 0.1, size = 0.5, alpha =0.4) + 
  scale_color_manual(values = c('#2e72af', 'grey')) +
  scale_y_reverse(breaks = seq(1,25,2)) +
  theme(strip.background = element_blank(),legend.position = 'none', 
        axis.text.x = element_text(angle = 45, hjust = 1), axis.title.x = element_blank()) +
  geom_hline(yintercept = 4.2, linetype = 'dashed') +
  ggtitle('300K-RNA')  +
  ylab('event rank')+
  theme(plot.margin = margin(0.5,0.5,0.5,0.5, "cm"))+
  annotate("text", x = 1, y = 2.5, label = paste('TP =', labs_sv[grepl('cryptic', splicing_event_class) & 
                                                                     !grepl('not', in_rna_studies) &
                                                                     rank_above_4 == 1 , n]), size = 3) +
  annotate("text", x = 1, y = 5.5, label = paste('FN =', labs_sv[grepl('cryptic', splicing_event_class) & 
                                                                     !grepl('not', in_rna_studies) &
                                                                     rank_above_4 == 0 , n]), size = 3)+
  annotate("text", x = 2, y = 2.5, label = paste('FP =',labs_sv[grepl('cryptic', splicing_event_class) & 
                                                                    grepl('not', in_rna_studies) &
                                                                    rank_above_4 == 1 , n]), size = 3)

G <- ggarrange(ca_sv, ca_sai, nrow = 1) %>% annotate_figure(fig.lab = 'Cryptic Activation', fig.lab.size = 12)


sv = jitter_data_sv %>%
  filter(grepl('cryptic', splicing_event_class)) %>%
  select(in_rna_studies, score = missplicing_event_rank_simple, meets_threshold = rank_above_4)%>%
  mutate(alg ='splicevault')

sai = jitter_data_sai %>%
  filter(grepl('cryptic', splicing_event_class))%>%
  select(in_rna_studies, score = delta, meets_threshold = delta_above_thresh)%>%
  mutate(alg ='spliceai')

addWorksheet(wb, "figure_4G")
writeData(wb, "figure_4G", as.data.frame(rbind(sv, sai)), startRow = 1, startCol = 1)
```


```{r}
saveWorkbook(wb, file = "../figs/source_data/source_data_fig_4.xlsx", overwrite = TRUE)
```

# putting panels together
```{r}
top = ggarrange(D,E,nrow = 1 , widths = c(1, 1.5))
bottom = ggarrange(panelF,G,nrow = 1)
ggarrange(top, bottom, ncol = 1)
ggsave('../figs/figure_four.pdf', height = 8, width = 9)
``` 

